File: backend/subscriptions-service/src/main/java/com/streaming/subscriptions/service/SubscriptionService.java
package com.streaming.subscriptions.service;

import com.streaming.common.event.ContentCreatedEvent;
import com.streaming.subscriptions.dto.SubscriptionRequest;
import com.streaming.subscriptions.model.UserSubscription;

import java.util.List;

public interface SubscriptionService {

    void subscribe(String userId, SubscriptionRequest request);

    void unsubscribe(String userId, String targetId);

    List<UserSubscription> getUserSubscriptions(String userId);

    void processContentEvent(ContentCreatedEvent event);
}

File: frontend/hear-we-go/src/app/services/subscription-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { Observable } from 'rxjs';

export interface SubRequest {
  targetId: string;
  targetName: string;
  type: 'ARTIST' | 'GENRE';
}

export interface UserSubscription {
  userId: string;
  targetId: string;
  targetName: string;
  type: string;
}

@Injectable({
  providedIn: 'root'
})
export class SubscriptionService {
  private apiUrl = `/api/subscriptions`;

  constructor(private http: HttpClient, private router: Router) {
  }

  subscribe(sub: SubRequest): Observable<void> {
    return this.http.post<void>(this.apiUrl, sub);
  }

  unsubscribe(targetId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${targetId}`);
  }

  getMySubscriptions(): Observable<UserSubscription[]> {
    return this.http.get<UserSubscription[]>(this.apiUrl);
  }
}
